fission-cli/library/Fission/CLI/Handler/Setup.hs

```diff
@@ -72,19 +79,33 @@ setup maybeOS maybeUsername maybeEmail = do
       Display.putOk "Done! You're all ready to go 🚀"
 
     Right () -> do
-      logUser @Text "🔑 Creating keys"
-      void . Key.Store.create $ Proxy @SigningKey
+      logUser @Text "🔑 Setting up keys"
       void . Key.Store.create $ Proxy @ExchangeKey
-
-      username <- do
-        Prompt.reaskYN "🏠 Do you have an existing account?" >>= \case
-          False ->
-            User.register maybeUsername maybeEmail
-
-          True -> do
-            logUser @Text "🔗 Please open auth.fission.codes on a signed-in device"
-            signingSK <- Key.Store.fetch $ Proxy @SigningKey
-            rootURL   <- getRemoteBaseUrl
-            Login.consume signingSK rootURL {baseUrlPath = "/user/link"} maybeUsername
-
-      Display.putOk $ "Done! Welcome to Fission, " <> textDisplay username <> " ✨"
+      case maybeKeyFile of
+        Just keyFile -> do
+          logDebug $ "🔑 Got a Keyfile: " <> keyFile
+          Key.Store.fromFile (Proxy @SigningKey) keyFile
+          attempt (sendAuthedRequest RootCredential User.whoami) >>= \case
+            Left err -> do
+              CLI.Error.put err "Invalid key file provided."
+              raise err
+            Right username -> do
+              baseURL <- getRemoteBaseUrl
+              signingPK  <- Key.Store.fetchPublic (Proxy @SigningKey)
+              _ <- WNFS.create (DID Key $ Ed25519PublicKey signingPK) "/"
+              Env.init username baseURL Nothing
+              Display.putOk $ "Done! Welcome to Fission, " <> textDisplay username <> " ✨"
+        Nothing -> do
+          void . Key.Store.create $ Proxy @SigningKey
+          username <- do
+            Prompt.reaskYN "🏠 Do you have an existing account?" >>= \case
+              False ->
+                User.register maybeUsername maybeEmail
+
+              True -> do
+                logUser @Text "🔗 Please open auth.fission.codes on a signed-in device"
+                signingSK <- Key.Store.fetch $ Proxy @SigningKey
+                rootURL   <- getRemoteBaseUrl
+                Login.consume signingSK rootURL {baseUrlPath = "/user/link"} maybeUsername
+
+          Display.putOk $ "Done! Welcome to Fission, " <> textDisplay username <> " ✨"
```

Comment:
Purely aesthetic suggestion:

This is readable, but I'd suggest breaking the `maybeKeyFile` cases into named helper functions. It just feels a bit long.